WordPress/gutenberg
# Notice: refactor to TypeScript

## What?
Refactor `Notice` component to TypeScript

Part of https://github.com/WordPress/gutenberg/issues/35744

## Why?

The refactor to TypeScript has many benefits (auto-generated docs, static linting and error prevention, better IDE experience). See https://github.com/WordPress/gutenberg/issues/35744 for more details

## How?
Followed the steps in the [TypeScript migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2022-11-29T14%3A45%3A34Z#refactoring-a-component-to-typescript)

## Testing Instructions
- Ensure all tests still pass
- Confirm Storybook examples and editor examples work as expected.

## Notes
- PR also includes some updates to `Snackbar` to re-use the new `Notice` types, as recommended in #45472
- Added some new stories to demonstrate actions, JSX children
- Included some in-line comments to add context where needed

## Followups
`OnRemove` and `OnDismiss` are confusing props. In `Notice` there's really no difference between them. In `Snackbar` however there _is_ a subtle difference in behavior, so it would be good to update the docs of that component to make the distinction more clear.
